caches/clan.go

```diff
@@ -0,0 +1,91 @@
+package caches
+
+import (
+	"fmt"
+	"math/rand"
+	"time"
+
+	gocache "github.com/patrickmn/go-cache"
+	"github.com/topfreegames/khan/models"
+)
+
+// ClansSummariesCache represents a cache for the RetrieveClansSummaries operation
+type ClansSummariesCache struct {
+	Cache                 *gocache.Cache
+	TimeToLive            time.Duration
+	TimeToLiveRandomError time.Duration
+}
+
+// GetClansSummaries returns a summary of the clans details for a given list of clans by their game
+// id and public ids
+func (c *ClansSummariesCache) GetClansSummaries(db models.DB, gameID string, publicIDs []string) ([]map[string]interface{}, error) {
+	resultMap := c.getCachedClansSummaries(gameID, publicIDs)
+	err := c.getAndCacheClansSummaries(db, gameID, resultMap)
+	if err != nil {
+		if _, ok := err.(*models.CouldNotFindAllClansError); !ok {
+			return nil, err
+		}
+	}
+	var result []map[string]interface{}
+	for _, publicID := range publicIDs {
+		if resultMap[publicID] != nil {
+			result = append(result, resultMap[publicID])
+		}
+	}
+	return result, err
+}
+
+func (c *ClansSummariesCache) getClanSummaryCacheKey(gameID, publicID string) string {
+	return fmt.Sprintf("%s/%s", gameID, publicID)
+}
+
+func (c *ClansSummariesCache) getClanSummaryCache(gameID, publicID string) map[string]interface{} {
+	clan, present := c.Cache.Get(c.getClanSummaryCacheKey(gameID, publicID))
+	if !present {
+		return nil
+	}
+	return clan.(map[string]interface{})
+}
+
+func (c *ClansSummariesCache) setClanSummaryCache(gameID, publicID string, clanPayload map[string]interface{}) {
+	ttl := c.TimeToLive - c.TimeToLiveRandomError
+	ttl += time.Duration(rand.Intn(int(2*c.TimeToLiveRandomError + 1)))
+	c.Cache.Set(c.getClanSummaryCacheKey(gameID, publicID), clanPayload, ttl)
+}
+
+func (c *ClansSummariesCache) getCachedClansSummaries(gameID string, publicIDs []string) map[string]map[string]interface{} {
```

Comment:
I thought about this you're suggesting at first, but I preferred the single responsibility principle in this case...

Reply:
The unified function is still clear its purpose, so imo it's fine. I think it would be easier to read and has the benefit of only iterating over the input once, so I'd take that over splitting into multiple functions.

Reply:
Ok you convinced me, big function in next commit!